### PR TITLE
Add a configuration item to choose the root file indicator.

### PR DESCRIPTION
**Pull request by shinyypig**
User can choose the root file indicator between `\documentclass` and `\begin{document}`.

Related to https://github.com/James-Yu/LaTeX-Workshop/issues/3804.

### CODE REVIEW

**Comment by James-Yu** [comment]
This PR needs discussion on its necessity before looking into the details. I don't think there is a use case where detecting `documentclass` won't work. The referenced issue need only to move `documentclass` to the said `main` file.

**Comment by shinyypig** [comment]
I believe that many users, including myself, use a "settings.tex" file to save the document format and packages used when writing multiple LaTeX projects. If we could use "\begin{document}" to detect the root file, then when we need to modify the style of these LaTeX projects, we only need to synchronize all "settings.tex" files. On the other hand, if we use "\documentclass" to detect the root file, we not only need to synchronize "settings.tex", but also need to modify the "\documentclass" of each LaTeX project.

**Comment by James-Yu** [comment]
I personally am not convinced by the use case. May I have your thoughts? @jlelong

**Comment by shinyypig** [comment]
This is a commonly used means of synchronizing format when writing papers with independent appendices and supplementary files.

**Comment by mihaiconstantin** [comment]
> I believe that many users, including myself, use a "settings.tex" file to save the document format and packages used when writing multiple LaTeX projects. If we could use "\begin{document}" to detect the root file, then when we need to modify the style of these LaTeX projects, we only need to synchronize all "settings.tex" files. On the other hand, if we use "\documentclass" to detect the root file, we not only need to synchronize "settings.tex", but also need to modify the "\documentclass" of each LaTeX project.

@shinyypig, I very much agree with this. Literally all my workflows follow the setup you indicated.

> I personally am not convinced by the use case. May I have your thoughts? @jlelong

@James-Yu, I am personally disappointed to read your comment. At the very least, please allow users to choose rather than forcing them into a use case.

**Comment by mihaiconstantin** [comment]
> I believe that many users, including myself, use a "settings.tex" file to save the document format and packages used when writing multiple LaTeX projects. If we could use "\begin{document}" to detect the root file, then when we need to modify the style of these LaTeX projects, we only need to synchronize all "settings.tex" files. On the other hand, if we use "\documentclass" to detect the root file, we not only need to synchronize "settings.tex", but also need to modify the "\documentclass" of each LaTeX project.

@shinyypig, I very much agree with this. Literally all my workflows follow the setup you indicated.

> I personally am not convinced by the use case. May I have your thoughts? @jlelong

@James-Yu, Since this change is likely to impact users' workflows, I think it would be greatly appreciated if you allowed users to choose rather than being force into a use case. What are your concerns regarding the ability to choose the root file?

**Comment by jlelong** [comment]
I have looked at various LaTeX editors and it clearly appears that setting the root file remains really challenging. When the editor tries to discover the root file by some internal mechanism, relying on the presence of `documentclass` seems much more common. This is how Overleaf proceeds for instance.

Even though I am not fully convinced by the use case,  I can understand that some projects are organised in a way that detecting `begin{document}` may be simpler. I have the impression that letting the user decide which string to detect does not harm.

File: package.json [hunk 1]
```diff
@@ -1151,6 +1151,16 @@
...
         },
+        "latex-workshop.latex.rootFile.indicator": {
+            "scope": "resource",
+            "type": "string",
+            "default": "\\documentclass",
+            "enum": [
+                "\\documentclass",
+                "\\begin{document}"
+            ],
+            "markdownDescription": "Determine the root file based on the specific indicator. If `\\documentclass`, the root file is the file containing `\\documentclass`. If `\\begin{document}`, the root file is the file containing `\\begin{document}`."
```
Comment (James-Yu) [inline package.json:1162]: The description can be more concise, e.g., Determine if the root file is detected based on the presence of `\\documentclass` or `\\begin{document}`.

Comment (James-Yu) [reply package.json:1162]: Also, the detection is not just `\\documentclass`, but `\\documentclass[]{}`.

File: src/components/manager.ts [hunk 2]
```diff
@@ -87,6 +85,27 @@ export class Manager {
...
+        const configuration = vscode.workspace.getConfiguration('latex-workshop')
+        const indicator = configuration.get('latex.rootFile.indicator')
+        let ROOT_INDICATOR = null
+        switch (indicator) {
+            case '\\documentclass':
+                ROOT_INDICATOR = /\\documentclass(?:\s*\[.*\])?\s*\{.*\}/ms
+                break
+            case '\\begin{document}':
+                ROOT_INDICATOR = /\\begin{document}/m
+                break
```
Comment (James-Yu) [inline src/components/manager.ts:101]: Do not use a CONSTANT variable in function. For this implementation, returning the corresponding regex is sufficient.

File: src/components/manager.ts [hunk 3]
```diff
@@ -87,6 +85,27 @@ export class Manager {
...
+            default:
+                ROOT_INDICATOR = /\\documentclass(?:\s*\[.*\])?\s*\{.*\}/ms
+                break
```
Comment (James-Yu) [inline src/components/manager.ts:104]: The default case should also log as an error.

File: src/components/manager.ts [hunk 4]
```diff
@@ -87,6 +85,27 @@ export class Manager {
         return path.normalize(out).split(path.sep).join('/')
     }
 
+    /**
+     * @returns the indicator if the root file
+     */
+    getRootIndictor() {
```
Comment (James-Yu) [inline src/components/manager.ts:91]: Typo. Also make this a proper [Typescript getter](https://www.typescripttutorial.net/typescript-tutorial/typescript-getters-setters/).

**Review by James-Yu — CHANGES_REQUESTED** [review]

File: src/components/manager.ts (hunk 2 above)
Comment (shinyypig) [reply src/components/manager.ts:101]: image

What should I use? Const or let?

File: src/components/manager.ts [hunk 5]
```diff
@@ -87,6 +85,23 @@ export class Manager {
...
+     * The indicator of the root file.
+     */
+    get rootIndictor() {
+        const configuration = vscode.workspace.getConfiguration('latex-workshop')
+        const indicator = configuration.get('latex.rootFile.indicator')
+        switch (indicator) {
+            case '\\documentclass[]{}':
+                return /\\documentclass(?:\s*\[.*\])?\s*\{.*\}/ms
+            case '\\begin{document}':
+                return /\\begin{document}/m
```
Comment (James-Yu) [inline src/components/manager.ts:98]: It's possible to have spaces or newlines before the left brace `{`. This case was neglected in previous versions. Please add this.

Comment (shinyypig) [reply src/components/manager.ts:98]: Sorry, I am not familiar with regex. These are directly copied from the source code. I don't know how to modify them. Could you please provide  me the required regex?

Comment (James-Yu) [reply src/components/manager.ts:98]: You may try /\\begin\s*{document}/m

**Pull request merged**